game-core/src/main/java/games/strategy/engine/data/GameParser.java

```diff
@@ -342,13 +349,23 @@ public void warning(final SAXParseException exception) {
       final boolean mustFind, final Function<String, T> function, final String errorName)
       throws GameParseException {
     final String name = element.getAttribute(attribute);
+    return getValidatedObject(name, mustFind, function, errorName);
+  }
+
+  private <T> T getValidatedObject(final String name, final boolean mustFind, final Function<String, T> function,
```

Comment:
Couple problems I can spot in this method:
- `function` is not a very a good parameter name. Does not describe what the function represents or does.
- API contains boolean parameter, at point of usage it's hard to understand (boolean parameters tend to indicate API problems as well, SRP violation)
- If `mustFind` is false, then we are not really saving any code, we can just call the `function` directly from the caller, why call this method?

Overall this method is taking the place of `Optional` API it seems and could be replaced according.

(1) whenever mustFind == false, then we can just invoke `function`, eg: `function.apply(name)`

(2) whenever mustFind == true, we can use Optional API, eg:
```
Optional.ofNullable(function.apply).orElseThrow(() -> new GameParseException(..));
```

Reply:
I'm sure this code could be refactored but was written this way to deal with the existing methods. Without increasing scope I don't think there is an easy win here.